crates/net/eth-wire/src/muxdemux.rs

```diff
@@ -0,0 +1,593 @@
+                if !mux_exhausted && self.inner.poll_ready_unpin(cx).is_ready() {
+                    continue
+                }
+                cx.waker().wake_by_ref();
+                // flush before returning pending
+                _ = self.inner.poll_flush_unpin(cx)?;
+            }
+            let mut bytes = match ready!(res) {
+                Some(Ok(bytes)) => bytes,
+                Some(Err(err)) => {
+                    _ = self.inner.poll_flush_unpin(cx)?;
+                    return Poll::Ready(Some(Err(err.into())))
+                }
+                None => {
+                    _ = self.inner.poll_flush_unpin(cx)?;
+                    return Poll::Ready(None)
+                }
+            };
+
+            // normalize message id suffix for capability
+            let cap = self.unmask_msg_id(&mut bytes[0])?;
+
+            // yield message for main stream
+            if *cap == self.owner {
+                _ = self.inner.poll_flush_unpin(cx)?;
+                return Poll::Ready(Some(Ok(bytes)))
+            }
+
+            // delegate message for stream clone
+            let tx = self.demux.get(cap).ok_or(CapabilityNotConfigured)?;
+            tx.send(bytes).map_err(|_| SendIngressBytesFailed)?;
+        }
+    }
+}
+
+impl<S, E> Sink<Bytes> for MuxDemuxStream<S>
+where
+    S: Sink<Bytes, Error = E> + CanDisconnect<Bytes> + Unpin,
+    MuxDemuxError: From<E>,
+{
+    type Error = MuxDemuxError;
+
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready_unpin(cx).map_err(Into::into)
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: Bytes) -> Result<(), Self::Error> {
+        let item = self.mask_msg_id(item);
+        self.inner.start_send_unpin(item).map_err(|e| e.into())
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_flush_unpin(cx).map_err(Into::into)
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        while let Ok(item) = self.mux.try_recv() {
+            self.inner.start_send_unpin(item)?;
+        }
+        _ = self.inner.poll_flush_unpin(cx)?;
+
+        self.inner.poll_close_unpin(cx).map_err(Into::into)
+    }
+}
+
+#[async_trait::async_trait]
+impl<S, E> CanDisconnect<Bytes> for MuxDemuxStream<S>
+where
+    S: Sink<Bytes, Error = E> + CanDisconnect<Bytes> + Unpin + Send + Sync,
+    MuxDemuxError: From<E>,
+{
+    async fn disconnect(&mut self, reason: DisconnectReason) -> Result<(), MuxDemuxError> {
+        if self.can_drop() {
+            return self.inner.disconnect(reason).await.map_err(Into::into)
+        }
+        Err(StreamInUse)
+    }
+}
+
+/// More or less a weak clone of the stream wrapped in [`MuxDemuxer`] but the bytes belonging to
+/// other capabilities have been filtered out.
+#[derive(Debug)]
+pub struct StreamClone {
```

Comment:
unsure about the naming yet,

I believe you had `WeakStream` or something.
perhaps `WeakCapabilityStream`?

Reply:
`WeakStreamClone` initially, 'weak' since it is some kind of reference to the stream but doesn't own it. if not stream clone then the naming I like the most is `ProtocolProxy` from _multiplex.rs_. `MuxDemuxer` or `MuxDemuxStream` is then the proxy server.